drivers/SmartThings/zigbee-window-treatment/src/aqara/aqara_utils.lua

```diff
@@ -0,0 +1,212 @@
+  if shadeLevel == 100 then
+    device:emit_event(capabilities.windowShade.windowShade.open())
+  elseif shadeLevel == 0 then
+    device:emit_event(capabilities.windowShade.windowShade.closed())
+  else
+    device:emit_event(capabilities.windowShade.windowShade.partially_open())
+  end
+end
+
+local function emit_shade_level_event(device, level)
+  device:emit_event(capabilities.windowShadeLevel.shadeLevel(level))
+end
+
+local function read_shade_position_attribute(device)
+  device:send(AnalogOutput.attributes.PresentValue:read(device))
+end
+
+local function shade_level_cmd(driver, device, command)
+  local level = command.args.shadeLevel
+  if level > 100 then
+    level = 100
+  end
+  level = utils.round(level)
+
+  -- update ui to the new level
+  emit_shade_level_event(device, level)
+
+  -- send
+  send_lift_percentage_cmd(device, command, level)
+end
+
+local function shade_open_cmd(driver, device, command)
+  send_lift_percentage_cmd(device, command, 100)
+end
+
+local function shade_close_cmd(driver, device, command)
+  send_lift_percentage_cmd(device, command, 0)
+end
+
+local function shade_pause_cmd(driver, device, command)
+  device:send_to_component(command.component, WindowCovering.server.commands.Stop(device))
+end
+
+local function setInitializedStateField(device, value)
+  device:set_field(INITIALIZED_STATE, value, { persist = true })
+end
+
+local function isInitializedStateField(device)
+  local initialized = device:get_field(INITIALIZED_STATE) or 0
+  return initialized == 1
+end
+
+local function shade_state_changed(device, value)
+  local state = value.value
+
+  -- update state ui
+  if state == SHADE_STATE_STOP then
+    -- local lastLevel = device:get_latest_state("main", capabilities.windowShadeLevel.ID,
+    --   capabilities.windowShadeLevel.shadeLevel.NAME) or 0
+    -- emit_shade_state_event(device, lastLevel)
+
+    -- read shade position to update the UI
+    read_shade_position_attribute(device)
+  elseif state == SHADE_STATE_OPEN then
+    device:emit_event(capabilities.windowShade.windowShade.opening())
+  elseif state == SHADE_STATE_CLOSE then
+    device:emit_event(capabilities.windowShade.windowShade.closing())
+  end
+end
+
+local function shade_position_changed(device, value)
+  local level = value.value
+  if level > 100 then
+    level = 100
+  end
+  level = utils.round(level)
+
+  -- update level ui
+  emit_shade_level_event(device, level)
+
+  -- local lastState = device:get_latest_state("main", capabilities.windowShade.ID,
+  --   capabilities.windowShade.windowShade.NAME) or 0
+  -- if lastState == "opening" then
+  -- elseif lastState == "closing" then
+  -- else
+  --   emit_shade_state_event(device, level)
+  -- end
+end
+
+local function is_matched_profile(device, profile)
+  for _, fingerprint in pairs(FINGERPRINTS) do
+    if device:get_model() == fingerprint.model and fingerprint.device_profile == profile then
+      return true
+    end
```

Comment:
This function is confusing for me. What's the advantage of the "profile" field over just matching on the model?

Reply:
I'll edit and push.